### PR TITLE
Add an option to automatically replace failed processes

### LINKED ISSUE #388: Failing reconciliation when there are processes not reporting to the cluster
If a pod is considered healthy, but there is a some other problem inside the container, we can see processes fail to connect to the cluster. When this happens, we will add them into the MissingProcesses list in the status. I think that if there is anything in this list, we should consider reconciliation incomplete.

### LINKED ISSUE #439: Option to automatically replace a process that has been missing for a certain period of time
As part of the project on automatic replacements, we should have a feature to automatically replace missing processes. When a process has been detected as missing for more than the configured time window, it should be added to the removal list in the status, and the operator should proceed with the normal replacement. This feature should be optional, and disabled by default. The default replacement threshold should be 30 minutes.

Reference for design: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/design/automatic_replacements.md

### PR DESCRIPTION
**Pull request by brownleej**
Resolves #439 
Resolves #388

### CODE REVIEW

File: controllers/update_status.go [hunk 1]
```diff
@@ -135,6 +134,7 @@ func (s UpdateStatus) Reconcile(r *FoundationDBClusterReconciler, context ctx.Co
 	if err != nil {
 		return false, err
 	}
+	removeDuplicateConditions(status)
```
Comment (johscheuer) [inline controllers/update_status.go:137]: Do you know why these duplicates appear? From the update function we actually shouldn't have any duplicates?

**Review by johscheuer — CHANGES_REQUESTED** [review]
A few comments otherwise it looks good overall.

File: api/v1beta1/foundationdbcluster_types.go [hunk 2]
```diff
@@ -453,13 +464,11 @@ func (processGroupStatus *ProcessGroupStatus) AddCondition(oldProcessGroups []*P
 	// Check if we got a condition for the condition type for the ProcessGroupStatus
 	if oldProcessGroupStatus != nil {
 		for _, condition := range oldProcessGroupStatus.ProcessGroupConditions {
-			if condition.ProcessGroupConditionType != conditionType {
-				continue
+			if condition.ProcessGroupConditionType == conditionType {
```
[the remark this answers is not in the document]
Comment (brownleej) [reply api/v1beta1/foundationdbcluster_types.go:467]: I think I was originally making more changes to this section of the code, then rolled that back. As it stands, this form is a couple lines shorter, so I'm happy with keeping it this way 😄

File: controllers/update_status.go (hunk 1 above)
Comment (brownleej) [reply controllers/update_status.go:137]: I don't, but I think it's valid to ensure explicitly that we are maintaining a contract of one entry per condition type.

File: controllers/pod_models.go [hunk 3]
```diff
@@ -1191,6 +1191,16 @@ func NormalizeClusterSpec(spec *fdbtypes.FoundationDBClusterSpec, options Deprec
 			source := fdbtypes.PublicIPSourcePod
 			spec.Services.PublicIPSource = &source
 		}
+
```
[the remark this answers is not in the document]
Comment (brownleej) [reply controllers/pod_models.go:1194]: You're right, that is a little weird.

File: controllers/replace_failed_pods.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+// ReplaceFailedPods identifies processes that have failed and need to be
+// replaced.
+type ReplaceFailedPods struct{}
+
+// Reconcile runs the reconciler's work.
+func (c ReplaceFailedPods) Reconcile(r *FoundationDBClusterReconciler, context ctx.Context, cluster *fdbtypes.FoundationDBCluster) (bool, error) {
+	hasNewRemovals := false
+
+	for _, processGroupStatus := range cluster.Status.ProcessGroups {
```
[the remark this answers is not in the document]
Comment (brownleej) [reply controllers/replace_failed_pods.go:38]: Yeah, I think we can factor this out and test it explicitly. There's another PR with big changes to the test suite, and once that's merged I think we can explore testing the individual subreconcilers more and relying less on testing the whole reconciliation.

File: controllers/cluster_controller_test.go [hunk 5]
```diff
@@ -838,6 +837,39 @@ var _ = Describe(fdbtypes.ProcessClassClusterController, func() {
...
+				}).Should(Equal(0))
+
+				Eventually(func() (int, error) {
+					err := k8sClient.List(context.TODO(), pods, getListOptions(cluster)...)
+					return len(pods.Items), err
+				}).Should(Equal(17))
+
+				sortPodsByID(pods)
+
+				log.Info("JPB checking pod names")
```
[the remark this answers is not in the document]
Comment (brownleej) [reply controllers/cluster_controller_test.go:865]: 😳

**Review by johscheuer — APPROVED** [review]
LGTM 👍

**Pull request merged**